### ISSUE TITLE
Numerous assert_malformed failures in test suite output in CI

### ISSUE THREAD

**Issue by peterhuene** [bytecodealliance/wasmtime#492]
#### Repro steps

1. `cargo test --all --exclude lightbeam --exclude wasmtime-wasi-c --exclude wasmtime-py -- --nocapture`

#### Expected results

The Wasm test suite tests should all pass.

#### Actual results

The Wasm test suites do pass, but most of the tests that check for malformed modules print error messages resulting from mismatched error strings (but otherwise pass).

In addition to looking like errors are being ignored by a non-failing assertion, the verbosity of these messages makes it hard to see which test suite tests ran in CI build logs.

**Comment by alexcrichton** [comment]
FWIW the tests are actually passing here, but tests are run [with `--nocapture`](https://github.com/bytecodealliance/wasmtime/blob/f592811c9aac88e589eab207a6c63719423ffe01/.github/workflows/main.yml#L170) (I'm not entirely sure why myself) and [neither `assert_invalid` nor `assert_malformed`](https://github.com/bytecodealliance/wasmtime/blob/f592811c9aac88e589eab207a6c63719423ffe01/crates/wast/src/wast.rs#L285-L328) will assert on their contents, they just print informational warnings. I think this is due to the fact that wasmtime's error messages are different than that of the spec test suite, and aligning them probably isn't exactly worth a lot of effort. 

I'd be personally in favor of just removing `--nocapture` or removing the debugging print altogether.

**Comment by peterhuene** [comment]
`--nocapture` is useful for `test-programs` tests as when they fail due to a panic it'll go to the output and then a more generic "test program returned non-zero" failure will fail the generated test.

I agree that we could probably get rid of the debug print from the `assert_*` commands and just leave them as no-ops until such time as we can properly assert on matching error messages.

**Comment by alexcrichton** [comment]
If a test fails though it should always print its own stdout/stderr in an isolated area dedicated to just that one test (Rust's test harness captures stdout/stderr by default), but if a test segfaults then `--nocapture` is the only way to see the output. In that sense I think we could probably still get by removing `--nocapture`?

**Comment by peterhuene** [comment]
Perhaps I'm not recalling correctly then, as I seem to remember having to use `--nocapture` to figure out what a test-program was really failing with 🤷‍♂ .

If that's the case then I'm totally 👍 on removing `--nocapture`.

**Comment by peterhuene** [comment]
FYI, my memory was indeed faulty and you're correct.  We don't need it for test programs.

**Comment by fitzgen** [comment]
Ok so I've got all the `assert_invalid` tests passing in `wasmparser`, so we can handle these in `wasmtime` with an update to our `wasmparser` dep and a little changes to our test generation in `build.rs`. (https://github.com/bytecodealliance/wasmparser/pull/186)

I've been poking at `assert_malformed` and this is a bit hairier. Basically, if a parser doesn't look *exactly* like the spec interpreter's parser, then getting all of these the same is going to be very difficult. There are lots of examples of things like hitting an EOF when parsing a value type and do you report that you have an invalid value type or that you hit an EOF? The spec chooses one or the other when neither is more correct (and sometimes it switches between them in seemingly arbitrary ways, that only make sense when you read its code!) and matching its behavior can be hard based on how the parser is structured and what kinds of helper functions you have. Also there are a bunch of things that I think are probably "bugs" in the spec interpreter's parser's error messages, where they report one thing but probably should report another.

Long story short, I don't think it is worth trying to match `assert_malformed` error messages. Instead, we should stop feeling guilty about not matching them, remove `TODO` comments in our `build.rs`, and stop printing these warnings in our tests. I'll start taking this course of action.

**Comment by alexcrichton** [comment]
@fitzgen I would agree that `wasmtime` here should skip the `assert_malformed` test error messages. That's technically covered by the `wat` crate anyway since it's only testing `*.wat` parse errors, and afaik it has nothing to do with the binary format.

The `wat` crate already validates all of its error messages against the test suite's `assert_malformed` messages as well. It has a [custom function](https://github.com/bytecodealliance/wat/blob/c6bad21517a7658940463b40883fb83e0c918250/tests/wabt.rs#L197-L216) to "adapt" from the expected error message to the actual error message. It's a pretty ugly function though and would be a bummer to duplicate here. (and probably not worth it)

**Comment by peterhuene** [comment]
This course of action sounds reasonable to me 👍.

**Comment by fitzgen** [comment]
> That's technically covered by the `wat` crate anyway since it's only testing `*.wat` parse errors, and afaik it has nothing to do with the binary format.

Not quite right -- it constructs a bunch of wasm binaries that have various syntax errors (as opposed to validation errors, which `assert_invalid` covers) and then checks that parsing these binaries produces the expected error message. For example, it produces binaries that have LEB128s that don't satisfy the max length is ceil(N / 7) property, etc.

I don't think this is something that `wat` can cover, so I'm a little confused at what it does here.

**Comment by alexcrichton** [comment]
Oh right sorry, I should read the code more closely. The `wat` crate [tests all "quote"](https://github.com/bytecodealliance/wat/blob/c6bad21517a7658940463b40883fb83e0c918250/tests/wabt.rs#L128-L132) modules, but not those that are likely `module binary "..."` which would do things like leb128 encodings and such. 

It may be the case though that skipping `module quote` tests (and leaving those to `wat`) may mean that the set of error messages we'd need to tweak and/or align would be manageable enough?

In any case starting out with `assert_invalid` sounds good, and we can tackle `assert_malformed` later.

**Comment by fitzgen** [comment]
> and we can tackle `assert_malformed` later.

I don't think it is worth tackling these later though. After poking at this for a bit, I don't think we will get anything of value out of it, and it would just be busy work.

**Comment by alexcrichton** [comment]
Oh right, sorry, I'm fine deferring to your judgment on that!

**Comment by tschneidereit** [comment]
> I don't think it is worth tackling these later though. After poking at this for a bit, I don't think we will get anything of value out of it, and it would just be busy work.

I mostly agree in that I think your argument for not fixing things on our end is compelling. At the same time, I'm a bit concerned about the risk of accidentally papering over real bugs here.

Would an alternative be to change the tests? Having spec tests that can't reasonably be tested against different implementations doesn't seem great. I remember similar issues with tc39's test262 suite, and IIRC the upstream tests were changed to have less implementation-specific behavior. (The difference is, of course, that tc39 doesn't maintain a spec interpreter. Given that the Wasm WG's spec interpreter is [non-normative](https://www.w3.org/2017/08/wasm-charter#deliverables), that doesn't seem to change the picture.)

**Comment by fitzgen** [comment]
To be clear, we _are_ testing (here and in `wasmparser`) that we never successfully parse a malformed module, we just aren't testing that the error messages are the same.

FWIW, I got through maybe a fifth of the test suite's `assert_malformed`, and did not discover any bugs.

**Comment by tschneidereit** [comment]
> To be clear, we _are_ testing (here and in `wasmparser`) that we never successfully parse a malformed module, we just aren't testing that the error messages are the same.

Understood! And I know that you'd not be in any way cavalier about us passing tests we really should be failing :) I'm slightly concerned about whether there's a risk of us really getting things wrong and that not being caught, but I admit that the circumstances under which that could happen are quite unlikely: AFAICT it'd require that all tests that'd cover a certain type of malformed content to *also* fail because of another type of malformed content. This is very different from typical JS tests, where you often have a `TypeError` and a description, and the `TypeError` really could be about basically anything :)

> FWIW, I got through maybe a fifth of the test suite's `assert_malformed`, and did not discover any bugs.

That, together with the above-described unlikeliness of anything concerning being papered over convinces me that there's nothing to do here once you've landed the patch to ignore the error messages. Thank you for the thorough explanation!
